src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -71,63 +71,200 @@ public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref
+            nint minLength = Math.Min(firstLength, secondLength);
+            nint offset = 0;
 
-            if ((byte*)minLength >= (byte*)(sizeof(UIntPtr) / sizeof(char)))
+            if (Sse2.IsSupported && minLength >= Vector128<ushort>.Count)
             {
-                if (Vector.IsHardwareAccelerated && (byte*)minLength >= (byte*)Vector<ushort>.Count)
+                // >= Sse2 intrinsics are supported, and length is enough to use them so use that path.
+                // We jump forward to the intrinsics at the end of them method so a naive branch predict
+                // will choose the non-intrinsic path so short lengths which don't gain anything aren't
+                // overly disadvantaged by having to jump over a lot of code. Whereas the longer lengths
+                // more than make this back from the intrinsics.
+                goto IntrinsicsCompare;
+            }
+
+            // Only use the Vector path if intrinsics aren't supported.
+            if (!Sse2.IsSupported && Vector.IsHardwareAccelerated)
+            {
+                nint lengthToExamine = minLength - Vector<ushort>.Count;
+                if (lengthToExamine >= 0)
                 {
-                    IntPtr nLength = minLength - Vector<ushort>.Count;
                     do
                     {
-                        if (Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) !=
-                            Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
+                        var matches = Vector.Equals(LoadVector(ref first, offset), LoadVector(ref second, offset));
+                        if (Vector<ushort>.Zero.Equals(matches))
                         {
-                            break;
+                            offset += Vector<ushort>.Count;
+                            continue;
                         }
-                        i += Vector<ushort>.Count;
-                    }
-                    while ((byte*)nLength >= (byte*)i);
-                }
 
-                while ((byte*)minLength >= (byte*)(i + sizeof(UIntPtr) / sizeof(char)))
-                {
-                    if (Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) !=
-                        Unsafe.ReadUnaligned<UIntPtr>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
-                    {
-                        break;
-                    }
-                    i += sizeof(UIntPtr) / sizeof(char);
+                        offset += LocateFirstFoundChar(matches);
+                        goto OffsetDifference;
+
+                    } while (lengthToExamine > offset);
                 }
             }
 
 #if BIT64
-            if ((byte*)minLength >= (byte*)(i + sizeof(int) / sizeof(char)))
+            ulong difference;
+            while (minLength >= (offset + sizeof(ulong) / sizeof(char)))
```

Comment:
This should read:

```cs
while (minLength - offset >= sizeof(ulong) / sizeof(char))
```

On 64-bit it won't overflow, but if this code is ever ported to a 32-bit code path it could integer overflow. So better to keep the code hygienic.

Reply:
Done

Reply:
Changed to the following as the Jit doesn't hoist the calc very well
```csharp
lengthToExamine = minLength - sizeof(ulong) / sizeof(char);
while (lengthToExamine > offset)
```